cmd/crates/soroban-spec-typescript/src/lib.rs

```diff
@@ -107,85 +145,86 @@ pub fn entry_to_ts(entry: &Entry) -> String {
                     .strip_prefix("Result<")
                     .unwrap()
                     .strip_suffix('>')
                     .unwrap()
                     .to_owned();
+                return_type = format!("Ok<{return_type}> | Err<Error_> | undefined");
             }
 
             let mut output = outputs
                 .get(0)
                 .map(|type_| {
                     if let Type::Custom { name } = type_ {
-                        format!("{name}FromXdr(response.xdr)")
+                        format!("{name}FromXdr(xdr)")
                     } else {
-                        format!("scValStrToJs(response.xdr) as {inner_return_type}")
+                        "scValStrToJs(xdr)".to_owned()
                     }
                 })
                 .unwrap_or_default();
             if is_result {
                 output = format!("new Ok({output})");
             }
-            let mut output = format!(
-                r#"
-    // @ts-ignore Type does exist
-    const response = await invoke(invokeArgs);
-    return {output};"#
-            );
+            if return_type != "void" {
+                output = format!(r#"return {output};"#);
+            };
             if is_result {
                 output = format!(
-                    r#"
-    try {{
-        {output}
-    }} catch (e) {{
-        //@ts-ignore
-        let err = getError(e.message);
-        if (err) {{
-            return err;
-        }} else {{
-            throw e;
-        }}
-    }}"#
+                    r#"try {{
+                {output}
+            }} catch (e) {{
+                //@ts-ignore
+                let err = getError(e.message);
+                if (err) {{
+                    return err;
+                }} else {{
+                    throw e;
+                }}
+            }}"#
                 );
             }
+            let parse_result_xdr = if return_type == "void" {
+                r#"parseResultXdr: () => {}"#.to_owned()
+            } else {
+                format!(
+                    r#"parseResultXdr: (xdr): {return_type} => {{
+            {output}
+        }}"#
+                )
+            };
+            let js_name = jsify_name(name);
+            let options = method_options(&return_type);
             let args = (!inputs.is_empty())
-                .then(|| format!("args: [{args}], "))
+                .then(|| format!("args: [{args}],\n        "))
                 .unwrap_or_default();
             format!(
-                r#"{ts_doc}export async function {name}({input} {{signAndSend, fee}}: {{signAndSend?: boolean, fee?: number}} = {{signAndSend: false, fee: 100}}){return_type} {{
-    let invokeArgs: InvokeArgs = {{
-        signAndSend,
-        fee,
-        method: '{name}', 
-        {args}
-    }};
-    {output}
+                r#"{ts_doc}export async function {js_name}<R extends ResponseTypes = undefined>({input}options: {options} = {{}}) {{
+    return await invoke({{
```

Comment:
can/should this return a `Promise`, instead? unless there's a usability reason to enforce this to be synchronous

Reply:
It does return a Promise, because it is `async`. 

If you don't await the `Promise` returned by `invoke`, you can end up with a double-wrapped Promise. At least, I've seen unexpected things happen, when I don't await inner calls like this.

Reply:
> be me
> js sdk maintainer for like 2 years
> still don't fully understand async/await semantics

thanks for elucidating :laughing: